sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+                .unwrap();
+            Self { value: bn_done }
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_mod_inv(
+                    lhs_ffi_in_addr: *const u64,
+                    rhs_ffi_in_addr: *const u64,
+                    bn_syscall_result: *mut u64,
+                ) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            let rhs_ffi_in = {
+                FfiBigNumber {
+                    data: modulus.value.as_ptr() as *const _ as u64,
+                    data_len: modulus.value.len() as u64,
+                    is_negative: modulus.is_negative,
+                }
+            };
+            let value_len = modulus.value.len();
+            let mut value = Vec::<u8>::with_capacity(value_len as usize);
+            let mut bn_syscall_result = {
+                FfiBigNumber {
+                    data: value.as_mut_ptr() as *mut _ as u64,
+                    data_len: value_len as u64,
+                    is_negative: false,
+                }
+            };
+            unsafe {
+                sol_bignum_mod_inv(
+                    &lhs_ffi_in as *const _ as *const u64,
+                    &rhs_ffi_in as *const _ as *const u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    value,
+                }
+            }
+        }
+    }
+
+    /// Log a `BigNum` from a program to the solana logs
+    pub fn log(&self) {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            crate::program_stubs::sol_log(&self.value.to_dec_str().unwrap());
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_log(lhs_ffi_in_addr: *const u64) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            unsafe { sol_bignum_log(&lhs_ffi_in as *const _ as *const u64) };
+        }
+    }
+}
+
+impl PartialEq for BigNumber {
+    fn eq(&self, other: &Self) -> bool {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.value == other.value && self.value.is_negative() == other.value.is_negative()
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value == other.value && self.is_negative == other.is_negative
```

Comment:
if you add `#[derive(PartialEq)]` to Self's defintion you can probably just do `self == other`

Reply:
Or better yet, get rid of this entirely because they can do `self == other` themselves :-)

Reply:
@jackcmay will change

Reply:
@jackcmay When I do that I get `error[E0369]: binary operation `==` cannot be applied to type `BigNumber``

Reply:
How about `#[derive(Eq)]`?

Reply:
@jackmay I have Eq and if I comment out the `impl` I get the following error:
```
can't compare `bignum::BigNumber` with `bignum::BigNumber`
the trait `std::cmp::PartialEq` is not implemented for `bignum::BigNumber`rustc(E0277)
```

Reply:
If I add `Eq` and comment out the `impl`, when running the BPF test I get:
```
 error[E0369]: binary operation `==` cannot be applied to type `BigNumber`
    --> rust/bignum/src/lib.rs:34:5
     |
  34 |     assert_eq!(base_bn_0, default_0);
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |     |
     |     BigNumber
     |     BigNumber
     |
     = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Reply:
How about this:
```rust
#[cfg(not(target_arch = "bpf"))]
#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, AbiExample)]

pub struct BigNumber {
    value: openssl::bn::BigNum,
}
#[cfg(target_arch = "bpf")]
#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, AbiExample)]
pub struct BigNumber {
    value: Vec<u8>,
    is_negative: bool,
}
```

Reply:
Will do